homeassistant/components/vodafone_station/coordinator.py

```diff
@@ -0,0 +1,164 @@
+        self._connected = False
+        self._connection_type: str | None = None
+        self._ip_address: str | None = None
+        self._last_activity: datetime | None = None
+        self._mac = mac
+        self._name = name
+        self._wifi: str | None = None
+
+    def update(
+        self,
+        dev_info: VodafoneStationDevice,
+        consider_home: float = DEFAULT_CONSIDER_HOME.total_seconds(),
+    ) -> None:
+        """Update device info."""
+        utc_point_in_time = dt_util.utcnow()
+
+        if self._last_activity:
+            consider_home_evaluated = (
+                utc_point_in_time - self._last_activity
+            ).total_seconds() < consider_home
+        else:
+            consider_home_evaluated = dev_info.connected
+
+        if not self._name:
+            self._name = dev_info.name or self._mac.replace(":", "_")
+
+        self._connected = dev_info.connected or consider_home_evaluated
+
+        if dev_info.connected:
+            self._last_activity = utc_point_in_time
+
+        self._connection_type = dev_info.connection_type
+        self._ip_address = dev_info.ip_address
+        self._wifi = dev_info.wifi
+
+    @property
+    def connection_type(self) -> str:
+        """Return connected status."""
+        return self._connection_type or ""
+
+    @property
+    def is_connected(self) -> bool:
+        """Return connected status."""
+        return self._connected
+
+    @property
+    def mac_address(self) -> str:
+        """Get MAC address."""
+        return self._mac
+
+    @property
+    def hostname(self) -> str | None:
+        """Get Name."""
+        return self._name
+
+    @property
+    def ip_address(self) -> str | None:
+        """Get IP address."""
+        return self._ip_address
+
+    @property
+    def last_activity(self) -> datetime | None:
+        """Return device last activity."""
+        return self._last_activity
+
+    @property
+    def wifi(self) -> str | None:
+        """Return device WIFi connection."""
+        return self._wifi
+
+
+class VodafoneStationRouter(DataUpdateCoordinator):
+    """Queries router running Vodafone Station firmware."""
+
+    def __init__(
+        self, host: str, ssl: bool, username: str, password: str, hass: HomeAssistant
+    ) -> None:
+        """Initialize the scanner."""
+
+        self._host = host
+        self._devices: dict[str, VodafoneStationDeviceInfo] = {}
+        self._data: dict[str, str] = {}
+        self._on_close: list[Callable] = []
+
+        self.api = VodafoneStationApi(host, ssl, username, password)
+
+        super().__init__(
+            hass=hass,
+            logger=_LOGGER,
+            name=f"{DOMAIN}-{host}-coordinator",
+            update_interval=timedelta(seconds=30),
+        )
+
+    @callback
+    def async_listener(self) -> None:
+        """Fetch data from the router."""
+
+        asyncio.run_coroutine_threadsafe(self._async_update_data(), self.hass.loop)
```

Comment:
```diff
diff --git a/homeassistant/components/vodafone_station/coordinator.py b/homeassistant/components/vodafone_station/coordinator.py
index a55cf9ff062..95d783e8e76 100644
--- a/homeassistant/components/vodafone_station/coordinator.py
+++ b/homeassistant/components/vodafone_station/coordinator.py
@@ -116,8 +116,24 @@ class VodafoneStationRouter(DataUpdateCoordinator):
     @callback
     def async_listener(self) -> None:
         """Fetch data from the router."""
-
-        asyncio.run_coroutine_threadsafe(self._async_update_data(), self.hass.loop)
+        dev_info: VodafoneStationDevice
+        for _, dev_info in self._devices.items():
+            mac = dev_info.mac
+            if IS_NEW:
+                dev = VodafoneStationDeviceInfo(mac, dev_info.name)
+                dev.update(dev_info)
+                self._devices[mac] = dev
+                async_dispatcher_send(
+                    self.hass, self.signal_device_new(mac), dev
+                )
+             else:
+                dev = self._devices[mac]
+                dev.update(dev_info)
+                # Don't send the signal if nothing has changed to avoid
+                # unnecessary state machine changes.
+                async_dispatcher_send(
+                    self.hass, self.signal_device_update(mac), dev
+                )
             
     @callback
     def async_on_close(self, func: CALLBACK_TYPE) -> None:
@@ -136,12 +152,7 @@ class VodafoneStationRouter(DataUpdateCoordinator):
         if not logged:
             raise ConfigEntryAuthFailed
 
-        devices = await self.api.get_all_devices()
-        dev_info: VodafoneStationDevice
-        for _, dev_info in devices.items():
-            dev = VodafoneStationDeviceInfo(dev_info.mac, dev_info.name)
-            dev.update(dev_info)
-            self._devices[dev_info.mac] = dev
+        self._devices = await self.api.get_all_devices()
         self._data = await self.api.get_user_data()
 
         await self.api.logout()
```

sample psuedo code

Reply:
I did a heavy clean-up of the code, now should be ok